### PR TITLE
feat: alllow specifying directives as bindings

### LINKED ISSUE #709: @Template.directives can take Bindings (not just types)
The `@Template.directives` should take full [`Binding`](https://github.com/angular/angular/blob/821f008462abdb24a18d4852bfe3fe04d8695823/modules/angular2/src/di/binding.js#L21) support.

```
@Template({
  url: '...',
  directives: [
    Directive1, 
    bind(Directive2).toClass(Directive2Impl),
  ]
}) 
```

## UPDATE 1

The reason for this issue is something like this.

```
<tabs>
  <pane></pane>
</tabs>
```

```
@Component({selector: 'tabs'})
class Tabs {}

@Component({selector: 'pane'})
class Pane {
  constructor(tabs:Tabs) {}
}
```

Here `Pane` can inject `Tabs`. But what if you would like to write a different tab which participates is this. Imagine this.

```
<view-stack>
  <pane></pane>
</view-stack>
```

```
@Component({selector: 'view-stack'})
class ViewStack extends Tabs {}
```

The above would not work even thought `ViewStack` implements `Tabs`.

So to fix this we would do this:

```
  directives: [
    Pane, 
    bind(Tabs).toClass(ViewStack),
  ]
```

And it would work as expected.

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Closes #709

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@tbosch would you mind reviewing this one for me?

File: modules/angular2/test/core/compiler/integration_spec.js [hunk 1]
```diff
@@ -275,6 +275,20 @@ export function main() {
...
+        }));
+
+        tb.createView(MyComp, {context: ctx}).then((view) => {
+          view.detectChanges();
+
+          expect(view.rootNodes).toHaveText('hello');
+          async.done();
+        });
+      }));
+
```
Comment (mhevery) [inline modules/angular2/test/core/compiler/integration_spec.js]: Can you add a tests where you create a binding such as: `bind(PublicAPI).toClass(PrivateImpl)` and then have a directive which injects 

```javascript
class SomeDecorator {
  constructor(@Parent() api:PublicAP) { 
    expect(api instanceof PrivateImpl).toBe(true);
  }
}
```

```html
<div public-api>
  <div child-decorator></div>
</div>
```

**Comment by pkozlowski-opensource** [comment]
@mhevery I've added [a test you've been mentioning](https://github.com/pkozlowski-opensource/angular/commit/444008ab1a2c5a9355de9cd657000501844712d1#diff-e1f3b6810121d09ca334907a177542fcR291) but while doing so I've realized that this is probably _not_ representing the use-case [described](https://github.com/angular/angular/issues/709#issuecomment-89316370) by @ajoslin . There are 2 differences I can see:
* here we are using meta-data (selectors etc.) from the "PublicAPI" while I think @ajoslin was expecting that a selector would be taken from the impl itself
* the essence of the trick we are playing here that we map a concrete impl (in terms of meta-data) to _one_ concrete implementation. This is a trick that we can play only once per template which probably is not enough for the more advanced use-cases

@ajoslin could you please verify if this commit fixes your use-case or not?

**Comment by tbosch** [comment]
@pkozlowski-opensource Can we merge this anyways? I need it for another issue...

**Comment by pkozlowski-opensource** [comment]
@tbosch merging this won't break anything so in this sense I think we can. But I still believe that this doesn't address the original issue and might be confusing (we read meta-data from one class and use a different class as impl, even if impl has different meta-data). 

I would say that we can merge this if we don't mind reviewing the original issue / introducing breaking changes in the near future.

**Comment by tbosch** [comment]
Ok, merged, but leaving the original issue open so that we don't forget...

**Pull request closed without merging**